sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc

```diff
@@ -461,51 +628,310 @@ is `name`.
+
+  namespace mykernels {
+
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclex::range_kernel<1>))
+  void bar(int *in, int *out) {/*...*/}
+
+  } // namespace mykernels
+)""";
+----
+
+The host code can compile this and get the kernel's `kernel` object like so:
+
+[source,c++]
+----
+sycl::kernel_bundle<sycl::bundle_state::ext_oneapi_source> kb_src = /*...*/;
+
+sycl::kernel_bundle<sycl::bundle_state::executable> kb = syclex::build(kb_src,
+  syclex::properties{syclex::registered_kernel_names{"mykernels::bar"}});
+
+sycl::kernel k = kb.ext_oneapi_get_kernel("mykernels::bar");
+----
+
+The {cpp} expression `"mykernels::bar"` computes the address of the kernel
+function `bar`.
+The host code then passes the same string (`"mykernels::bar"`) to
+`ext_oneapi_get_kernel` in order to get the `kernel` object.
+The string must have exactly the same content as the string that was used to
+construct the property, without even any whitespace differences.
+
+The application can also obtain the compiler-generated (i.e. mangled) name for
+the kernel by calling `ext_oneapi_get_raw_kernel_name` like this:
+
+[source,c++]
+----
+sycl::kernel_bundle<sycl::bundle_state::ext_oneapi_source> kb_src = /*...*/;
+
+sycl::kernel_bundle<sycl::bundle_state::executable> kb = syclex::build(kb_src,
+  syclex::properties{syclex::registered_kernel_names{"mykernels::bar"}});
+
+std::string mangled_name = kb.ext_oneapi_get_raw_kernel_name("mykernels::bar");
+----
+
+Again, the string passed to `ext_oneapi_get_raw_kernel_name` must have exactly
+the same content as the string that was used to construct the
+`registered_kernel_names` property.
+The application may also pass this compiler-generated (i.e. mangled) name to
+`ext_oneapi_get_kernel` in order to get the `kernel` object.
+
+==== Instantiating templated kernel functions
+
+The `registered_kernel_names` property can also be used to instantiate a
+kernel that is defined as a function template.
+For example, consider source code that defines a kernel function template like
+this:
+
+[source,c++]
+----
+std::string source = R"""(
+  #include <sycl/sycl.hpp>
+  namespace syclex = sycl::ext::oneapi::experimental;
+
+  template<typename T>
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclex::range_kernel<1>))
+  void bartmpl(T *in, T *out) {/*...*/}
+)""";
+----
+
+The application can use the `registered_kernel_names` property to instantiate
+the template for specific template arguments.
+For example, this host code instantiates the template twice and gets a `kernel`
+object for each instantiation:
+
+[source,c++]
+----
+sycl::kernel_bundle<sycl::bundle_state::ext_oneapi_source> kb_src = /*...*/;
+
+sycl::kernel_bundle<sycl::bundle_state::executable> kb = syclex::build(kb_src,
+  syclex::properties{syclex::registered_kernel_names{{"bartmpl<float>", "bartmpl<int>"}});
+
+sycl::kernel k_float = kb.ext_oneapi_get_kernel("bartmpl<float>");
+sycl::kernel k_int = kb.ext_oneapi_get_kernel("bartmpl<int>");
+----
```

Comment:
This is really cool. However, what if the user does not instantiate the template? What if they just query the name and try to use it? Would the following code be an error?
```c++
sycl::kernel_bundle<sycl::bundle_state::executable> kb = syclex::build(kb_src,
  syclex::properties{syclex::registered_kernel_names{{"bartmpl"}});

sycl::kernel k = kb.ext_oneapi_get_kernel("bartmpl");
```

Reply:
The section "Using the `registered_kernel_names` property" says:

> Each string in the property must be the C++ expression for a pointer to a kernel function.

Your example violates this requirement, so I think the behavior is undefined.  The string "bartmpl" is the name of a function template; it is not a valid expression for a pointer to a function.

I suspect our implementation will fail in the `build` call because it will try to compile "bartmpl" in a context where a function pointer is expected.

Reply:
Okey, perfect. Thanks, Greg!